torch/distributed/tensor/_api.py

```diff
@@ -212,6 +221,244 @@ def backward(ctx, grad_output: "DTensor"):  # type: ignore[override]
+        )
+
+    # normalize tensor dims in shard_order and convert str mesh dim into int
+    normalized_shard_order: dict[int, list[int]] = {}
+    has_stridedshard_in_placements = False
+
+    if shard_order is not None:
+        for tensor_dim, mesh_dims in shard_order.items():
+            if tensor_dim < -tensor_rank or tensor_dim >= tensor_rank:
+                raise ValueError(
+                    f"tensor dim {tensor_dim} specified in `shard_order` is out of range for tensor_rank {tensor_rank}."
+                )
+            tensor_dim = tensor_dim + tensor_rank if tensor_dim < 0 else tensor_dim
+            if tensor_dim in normalized_shard_order:
+                raise ValueError(
+                    f"both normalized tensor dim {tensor_dim} and un-normalized "
+                    f"tensor dim {tensor_dim - tensor_rank}) is specified in `shard_order`!"
+                )
+            normalized_shard_order[tensor_dim] = []
+            for mesh_dim in mesh_dims:
+                if isinstance(mesh_dim, str):
+                    mesh_dim = device_mesh._get_mesh_dim_by_name(mesh_dim)
+                if mesh_dim < 0 or mesh_dim >= device_mesh.ndim:
+                    raise IndexError(
+                        f"mesh dim {mesh_dim} specified in `shard_order` is out of range "
+                        f"for placements of length {device_mesh.ndim}"
+                    )
+                normalized_shard_order[tensor_dim].append(mesh_dim)
+
+    # set default placements to replicated if not specified
+    placement_tuple: tuple[Placement, ...]
+    if placements is None:
+        assert shard_order is not None
+        # convert from shard_order to placements
+        placement_tuple = _convert_shard_order_to_placements(
+            normalized_shard_order, device_mesh
+        )
+        shard_order_tuple = _from_dict_to_ShardOrder(normalized_shard_order)
+    else:
+        normalized_placements = list(placements)  # type: ignore[assignment]
+        for i, placement in enumerate(placements):
+            if placement.is_shard():
+                tensor_dim = placement.dim  # type: ignore[attr-defined]
+                if tensor_dim < -tensor_rank or tensor_dim >= tensor_rank:
+                    raise ValueError(
+                        f"tensor dim {tensor_dim} specified in `placements` is out of range for tensor_rank {tensor_rank}."
+                    )
+                tensor_dim = tensor_dim + tensor_rank if tensor_dim < 0 else tensor_dim
+
+                # reconstruct `placement` object in case it is `_StridedShard` for backward compatibility
+                if isinstance(placement, _StridedShard):
+                    has_stridedshard_in_placements = True
+                    normalized_placements[i] = _StridedShard(
+                        tensor_dim, split_factor=placement.split_factor
+                    )
+                else:
```

Comment:
i have a couple thoughts here

1- as mentioned above, we might have other types of shard placements (e.g. DynamicShard), and it'd be good to not have to change this code.
2- it's kind of annoying to normalize the tensor dim so late in the game.  I wonder if we can do it earlier.  The Shard constructor can't actually do it becuase it doesn't know anything about the tensor, right?  We could add a 'normalize(tensor_rank)' helper to the base 'Placement' class, and call that on the user-passed placements, which would avoid the normalization code being here and also ensure it works to normalize any kind of placement.